### DISCUSSION TITLE
How to handle a USB interface with a transfer type of "Interrupt"?

### DISCUSSION THREAD

**Discussion by SteveSpencer** [rsta2/circle#485] [Q&A] [answered]
My USB display/touchscreen has the following endpoints
```
     Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x01  EP 1 OUT
        bmAttributes            2
          Transfer Type            Bulk
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0040  1x 64 bytes
        bInterval               0
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0020  1x 32 bytes
```
Now, I understand how to use the `Transfer` method to send data out to the first endpoint, which is how the data is sent to the display.

If I  use `SubmitRequest` on the second endpoint, it will stall until data is available. Obviously, I can use `SubmitAsyncRequest` and a completion routine so that it does not block. Effectively, I am polling the device.

Here is where my ignorance becomes apparent. Is it possible to set an `interrupt handler` so that when data is available via the endpoint, a function can then submit a request to get it (and in my case, pass the data through to the CTouchScreenDevice) ?

If not, then I am left with polling the device (via SubmitAsyncRequest and a completion routine). What would be a suitable mechanism for this ? 
At the moment, I do a SubmitAsyncRequest, and in the completion routine, do another one after processing the newly arrived data, so effectively the code is continually requesting the touch status. I don't think this is necessarily a good thing to do, but I am not sure how else to do it.

**Comment by rsta2** [comment, marked as answer]
Have a look on how the [HID driver](https://github.com/rsta2/circle/blob/749fe99301846217439a40ac7bda547a986db0f7/lib/usb/usbhiddevice.cpp#L197) is doing this. In fact the device is polled by the USB driver in the lower layer on RPi 1-3, there is no other way. The completion routine is only called, when data arrives.

You can call the touchscreen [ReportHandler](https://github.com/rsta2/circle/blob/749fe99301846217439a40ac7bda547a986db0f7/include/circle/input/touchscreen.h#L61) from the USB completion routine. The USB HID touchscreen driver is doing this too, for example [here](https://github.com/rsta2/circle/blob/749fe99301846217439a40ac7bda547a986db0f7/lib/usb/usbtouchscreen.cpp#L185).

There is no other way to do this.

**Reply by SteveSpencer** [reply]
OK. So the term 'Interrupt' doesn't mean what I initially thought, but that is fine. 
My code already does those things so it should be fine.

I am having issues with the display side where I get black horizontal lines on the display, so I won't be submitting a PR for the driver and the LVGL integration until I have resolved that - I suspect it is to do with the fact that I am sending a rectangle of image data, and the y and/or height value may be slightly off.

Unfortunately the protocol spec I have is for a much earlier version of the device firmware, and RoboPeak haven't responded to my email, not suprisingly, since it is an 'abandoned' product.

**Reply by rsta2** [reply]
I think for the user it is an "interrupt" EP, because the completion routine is only called, when data is available. It is caused by the limitations of the USB host controller, that the EP has to be polled by the driver. This is not the case for RPi 4 and 5, where the xHCI controller is doing this on its own.

I think the Linux driver for this device is [here](https://github.com/robopeak/rpusbdisp/tree/master/drivers/linux-driver), but it's 10 years old.

**Reply by SteveSpencer** [reply]
Yes, that is the driver - and my device is the same age but still working.
The driver code doesn't match the protocol information I have, so I have done lots of experimenting to get it to work properly.

Apart from the one issue (which became apparent when I was using LVGL), things are working, and I managed to get the RLE compression for the display data working, so the amount of data being sent is significantly reduced in most cases. Once I figure out the problem I will submit a PR to provide driver code and my LVGL code.

I notice that the _old_ version of LVGL is in addons. I will look at how much work getting v9 to compile and run would be, but not right now!

**Reply by rsta2** [reply]
Please let me do the LVGL update by myself. It must be in sync with my MiniSynth Pi project. I have this already on my list.

**Reply by rsta2** [reply]
The LVGL support has been updated to v9.2 on the *develop* branch.

**Reply by SteveSpencer** [reply]
Wow. That was fast. I have an app that uses LVGL 9, which was running on an ESP32. Using Circle I can now move it to a Pi Zero W. My reason for wanting to change it was so I could edit configuration data on a uSD card prior to deployment. 
Although the h/w costs more, it gives me much more flexibility, which for non-commercial targets is a better fit.

Thanks.

**Reply by rsta2** [reply]
Actually I had expected more changes from LVGL 8.3.11 to 9.2.0, but luckily it wasn't that difficult. They renamed a lot of functions (e.g. `lv_image_` instead of `lv_img_`), but I think the old names still work for compatibility. Yes, the RPi Zero W is a great device for bare metal apps. You are welcome.
